Group Operator Count does not work as expected

I saw that group operator count was not documented.
Per comments it is expected to work this way : 
```
* Group:Number:COUNT(".") will count all items having a string state of one character
* Group:Number:COUNT("[5-9]") will count all items having a string state between 5 and 9
```

So I was about to give him some unit tests but I face an issue : 
- Per global logic (in ItemDTOMapper.mapFunction) group operators that accepts arguments are expected to return the same type than the operator but this does not work for Count as the operator is StringType while output is DecimalType.
I don't know how this happened to work on OH1 but ESH complains : 

Definition : 
```
Group:String:COUNT(".*") gSensorAirq	"Qualité Air [%d]"	<airquality>	(gSensors)


2017-10-27 08:00:24.128 [ERROR] [pse.smarthome.core.items.GenericItem] - Tried to set invalid state 5 (DecimalType) on item gSensorAirq of type StringItem, ignoring it
```
but value of the group seems good.

If I switch definition to Number : 
```
Group:Number:COUNT(".*") gSensorAirq	"Qualité Air [%d]"	<airquality>	(gSensors)

2017-10-27 09:45:05.068 [WARN ] [arthome.core.items.dto.ItemDTOMapper] - State '.*' is not valid for a group item with base type 'Number'
```

while changing label value format to %s :
```
Group:String:COUNT(".*") gSensorAirq	"Qualité Air [%s]"	<airquality>	(gSensors)
```
does not log error and does not work at all.

Has somebody an idea on how to address this ?

--- a/bundles/core/org.eclipse.smarthome.core.test/src/test/java/org/eclipse/smarthome/core/items/dto/ItemDTOMapperTest.java
+++ b/bundles/core/org.eclipse.smarthome.core.test/src/test/java/org/eclipse/smarthome/core/items/dto/ItemDTOMapperTest.java
@@ -11,12 +11,13 @@
  */
 package org.eclipse.smarthome.core.items.dto;
 
-import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.hamcrest.CoreMatchers.*;
 import static org.junit.Assert.assertThat;
 
 import org.eclipse.smarthome.core.items.GroupFunction;
 import org.eclipse.smarthome.core.library.items.NumberItem;
 import org.eclipse.smarthome.core.library.types.ArithmeticGroupFunction;
+import org.eclipse.smarthome.core.library.types.StringType;
 import org.junit.Test;
 
 public class ItemDTOMapperTest {
@@ -34,6 +35,9 @@ public class ItemDTOMapperTest {
         GroupFunction gFunc = ItemDTOMapper.mapFunction(item1, gFuncDTO);
 
         assertThat(gFunc, instanceOf(ArithmeticGroupFunction.Count.class));
+        assertThat(gFunc.getParameters().length, is(1));
+        assertThat(gFunc.getParameters()[0], instanceOf(StringType.class));
+
     }
 
 }

--- a/bundles/core/org.eclipse.smarthome.core.test/src/test/java/org/eclipse/smarthome/core/items/dto/ItemDTOMapperTest.java
+++ b/bundles/core/org.eclipse.smarthome.core.test/src/test/java/org/eclipse/smarthome/core/items/dto/ItemDTOMapperTest.java
@@ -1,0 +1,39 @@
+/**
+ * Copyright (c) 2014-2017 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+/**
+ * @author Stefan Triller - initial contribution
+ */
+package org.eclipse.smarthome.core.items.dto;
+
+import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.junit.Assert.assertThat;
+
+import org.eclipse.smarthome.core.items.GroupFunction;
+import org.eclipse.smarthome.core.library.items.NumberItem;
+import org.eclipse.smarthome.core.library.types.ArithmeticGroupFunction;
+import org.junit.Test;
+
+public class ItemDTOMapperTest {
+
+    @Test
+    public void testMapFunctionWithNumberItemAndCountFunction() {
+
+        // testing Group:Number:Count(".*hello.*")
+        NumberItem item1 = new NumberItem("item1");
+
+        GroupFunctionDTO gFuncDTO = new GroupFunctionDTO();
+        gFuncDTO.name = "COUNT";
+        gFuncDTO.params = new String[] { ".*hello.*" };
+
+        GroupFunction gFunc = ItemDTOMapper.mapFunction(item1, gFuncDTO);
+
+        assertThat(gFunc, instanceOf(ArithmeticGroupFunction.Count.class));
+    }
+
+}

--- a/bundles/core/org.eclipse.smarthome.core.test/src/test/java/org/eclipse/smarthome/core/library/types/ArithmeticGroupFunctionTest.java
+++ b/bundles/core/org.eclipse.smarthome.core.test/src/test/java/org/eclipse/smarthome/core/library/types/ArithmeticGroupFunctionTest.java
@@ -203,6 +203,18 @@ public class ArithmeticGroupFunctionTest {
         assertEquals(new DecimalType("234.95"), state);
     }
 
+    @Test
+    public void testCountFunction() {
+        items.add(new TestItem("TestItem1", new StringType("hello world")));
+        items.add(new TestItem("TestItem2", new StringType("world")));
+        items.add(new TestItem("TestItem3", new StringType("foo bar")));
+
+        function = new ArithmeticGroupFunction.Count(new StringType(".*world.*"));
+        State state = function.calculate(items);
+
+        assertEquals(new DecimalType("2"), state);
+    }
+
     class TestItem extends GenericItem {
 
         public TestItem(String name, State state) {

--- a/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/dto/ItemDTOMapper.java
+++ b/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/dto/ItemDTOMapper.java
@@ -18,6 +18,7 @@ import org.eclipse.smarthome.core.items.GroupItem;
 import org.eclipse.smarthome.core.items.Item;
 import org.eclipse.smarthome.core.items.ItemFactory;
 import org.eclipse.smarthome.core.library.types.ArithmeticGroupFunction;
+import org.eclipse.smarthome.core.library.types.StringType;
 import org.eclipse.smarthome.core.types.State;
 import org.eclipse.smarthome.core.types.TypeParser;
 import org.slf4j.LoggerFactory;
@@ -82,24 +83,11 @@ public class ItemDTOMapper {
 
     public static GroupFunction mapFunction(Item baseItem, GroupFunctionDTO function) {
         List<State> args = new ArrayList<State>();
-        if (function.params != null) {
-            for (String arg : function.params) {
-                State state = TypeParser.parseState(baseItem.getAcceptedDataTypes(), arg);
-                if (state == null) {
-                    LoggerFactory.getLogger(ItemDTOMapper.class).warn(
-                            "State '{}' is not valid for a group item with base type '{}'",
-                            new Object[] { arg, baseItem.getType() });
-                    args.clear();
-                    break;
-                } else {
-                    args.add(state);
-                }
-            }
-        }
 
         GroupFunction groupFunction = null;
         switch (function.name.toUpperCase()) {
             case "AND":
+                args = parseStates(baseItem, function.params);
                 if (args.size() == 2) {
                     groupFunction = new ArithmeticGroupFunction.And(args.get(0), args.get(1));
                 } else {
@@ -108,6 +96,7 @@ public class ItemDTOMapper {
                 }
                 break;
             case "OR":
+                args = parseStates(baseItem, function.params);
                 if (args.size() == 2) {
                     groupFunction = new ArithmeticGroupFunction.Or(args.get(0), args.get(1));
                 } else {
@@ -116,6 +105,7 @@ public class ItemDTOMapper {
                 }
                 break;
             case "NAND":
+                args = parseStates(baseItem, function.params);
                 if (args.size() == 2) {
                     groupFunction = new ArithmeticGroupFunction.NAnd(args.get(0), args.get(1));
                 } else {
@@ -124,6 +114,7 @@ public class ItemDTOMapper {
                 }
                 break;
             case "NOR":
+                args = parseStates(baseItem, function.params);
                 if (args.size() == 2) {
                     groupFunction = new ArithmeticGroupFunction.NOr(args.get(0), args.get(1));
                 } else {
@@ -132,8 +123,12 @@ public class ItemDTOMapper {
                 }
                 break;
             case "COUNT":
-                if (args.size() == 1) {
-                    groupFunction = new ArithmeticGroupFunction.Count(args.get(0));
+                State countParam = null;
+                if (function.params != null && function.params.length == 1) {
+                    countParam = new StringType(function.params[0]);
+                }
+                if (countParam != null) {
+                    groupFunction = new ArithmeticGroupFunction.Count(countParam);
                 } else {
                     LoggerFactory.getLogger(ItemDTOMapper.class)
                             .error("Group function 'COUNT' requires one argument. Using Equality instead.");
@@ -164,6 +159,28 @@ public class ItemDTOMapper {
         }
 
         return groupFunction;
+    }
+
+    private static List<State> parseStates(Item baseItem, String[] params) {
+        List<State> states = new ArrayList<State>();
+
+        if (params == null) {
+            return states;
+        }
+
+        for (String param : params) {
+            State state = TypeParser.parseState(baseItem.getAcceptedDataTypes(), param);
+            if (state == null) {
+                LoggerFactory.getLogger(ItemDTOMapper.class).warn(
+                        "State '{}' is not valid for a group item with base type '{}'",
+                        new Object[] { param, baseItem.getType() });
+                states.clear();
+                break;
+            } else {
+                states.add(state);
+            }
+        }
+        return states;
     }
 
     /**

--- a/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/dto/ItemDTOMapper.java
+++ b/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/dto/ItemDTOMapper.java
@@ -123,11 +123,8 @@ public class ItemDTOMapper {
                 }
                 break;
             case "COUNT":
-                State countParam = null;
                 if (function.params != null && function.params.length == 1) {
-                    countParam = new StringType(function.params[0]);
-                }
-                if (countParam != null) {
+                    State countParam = new StringType(function.params[0]);
                     groupFunction = new ArithmeticGroupFunction.Count(countParam);
                 } else {
                     LoggerFactory.getLogger(ItemDTOMapper.class)

